### PR TITLE
Avoid Sharing Parameter Default Values

### PR DESCRIPTION
**Pull request by theMomax**
# Avoid Sharing Parameter Default Values

## :recycle: Current situation
All Handlers use the same instance of a @Parameter's default value. This is no problem for value-types. The @Parameter's setter is not available to the user, thus value-types cannot be altered and sharing the default value is fine.

However, reference-types (i.e. class objects) can be altered nevertheless. So, when using a class type on an @Parameter with a default value, Handlers share this default value and can alter it.

A detailed example is shown in #122.

## :bulb: Proposed solution
The `Parameter.defaultValue` is stored as a closure. This closure is called each time a new `ConnectionContext` is created. 

### Problem that is solved
See #122

### Implications
The closure is captured using the `@autoclosure` mechanism, thus there are no changes to the public API. Internal changes should not affect exporters.

## :heavy_plus_sign: Additional Information
The same pattern used used in this PR will most likely also be used for implementing the `@State` property wrapper.

### Related PRs
None.

### Testing
Previous tests should avoid regression regarding a `Parameter`'s `options`. The new `testMutationOnClassTypeDefaultParameterIsNotShared` tests if the default value is really not shared between `Handler`.

### Reviewer Nudging
Should be rather easy. The interesting part regarding parameter options (necessity, nullability, ...) happens in `EndpointParameter.swift`. The unwrapping of the closure happens in `Validator.swift`. Note that some I had to add another seemingly redundant initializer to `Parameter`. This was required because otherwise the swift compiler would crash.

### CODE REVIEW

File: Tests/ApodiniTests/EndpointsTreeTests.swift [hunk 1]
```diff
@@ -48,7 +48,7 @@ final class EndpointsTreeTests: ApodiniTests {
             // swiftlint:disable:next force_unwrapping
             (0...times!)
                 .map { _ in
-                    "Hello \(name) born in \(birthdate.year)!"
+                    "Hello \(name) born in \("birthdate.year")!"
```
Comment (bauer-andreas) [inline Tests/ApodiniTests/EndpointsTreeTests.swift]: Those parentheses seem a bit odd to me? Is this intentional? Doesn't seem to impact anything but still.

File: Tests/ApodiniTests/ParameterMutabilityTests.swift [hunk 2]
```diff
@@ -28,6 +28,29 @@ class ParameterMutabilityTests: ApodiniTests {
...
+        var string: String
+        
+        init(string: String) {
+            self.string = string
+        }
+    }
+    
+    struct TestHandlerUsingClassType: Handler {
+        @Parameter
+        var name = StringClass(string: "Apodini")
```
Comment (bauer-andreas) [inline Tests/ApodiniTests/ParameterMutabilityTests.swift]: Would you mind adding a test case (or adding an assert to the test case) for a optional parameter with a default value.
As Parameters with optional Type are treated differently than parameters with a non optional type (and also the defaultValue closure is constructed differently) I would find a test case like described important.

**Review by bauer-andreas — APPROVED** [review]
Thanks for the quick fix of the bug. The solution looks good to me.
Just think we need a additional test case for Parameters with optional types, otherwise this is ready to merge from my side 👍

File: Tests/ApodiniTests/EndpointsTreeTests.swift (hunk 1 above)
Comment (theMomax) [reply Tests/ApodiniTests/EndpointsTreeTests.swift]: No, it isn't...I had to "uncomment" quite some code in order to find out why the compiler crashed and didn't catch this one when adding stuff back in. Thanks for the hint!

File: Tests/ApodiniTests/ParameterMutabilityTests.swift (hunk 2 above)
Comment (theMomax) [reply Tests/ApodiniTests/ParameterMutabilityTests.swift]: Yes, seems reasonable. I will add one.

**Review by PSchmiedmayer — COMMENTED** [review]
@theMomax I like the addition of capturing the default values using the auto closure 🚀
The only comment I would have would be to consider removing the `defaultClosurePresent` boolean flag and replace it with `defaultValue` being an optional itself (`(() -> Element)?`). 

I have experimented with this a bit while reviewing and testing your code and created a pull request to this branch to showcase the diff (#130).
What do you think about the idea? I think it would make the parameter implementation it bit clearer and would introduce a single source of truth for `defaultValue` and would enable us to put all the information into this single property. I reran the test suit and experimented a bit with the change and could not find any regressions.

**Comment by theMomax** [comment]
> @theMomax I like the addition of capturing the default values using the auto closure 🚀
> The only comment I would have would be to consider removing the `defaultClosurePresent` boolean flag and replace it with `defaultValue` being an optional itself (`(() -> Element)?`).
> 
> I have experimented with this a bit while reviewing and testing your code and created a pull request to this branch to showcase the diff (#130).
> What do you think about the idea? I think it would make the parameter implementation it bit clearer and would introduce a single source of truth for `defaultValue` and would enable us to put all the information into this single property. I reran the test suit and experimented a bit with the change and could not find any regressions.

@PSchmiedmayer there is one core difference between our approaches: Your implementation requires default values of parameters to be constant while mine doesn't. I have no strong opinion on what approach is more desirable. See my detailed comment [here](https://github.com/Apodini/Apodini/pull/130#discussion_r557220113).

**Comment by bauer-andreas** [comment]
> @PSchmiedmayer there is one core difference between our approaches: Your implementation requires default values of parameters to be constant while mine doesn't. I have no strong opinion on what approach is more desirable. See my detailed comment [here](https://github.com/Apodini/Apodini/pull/130#discussion_r557220113).

Not sure if I looked at the difference in implementation detailed enough, but wouldn't it be desirable that the default value does not change? In the case where a description of an WebService like OpenAPI creates it, include default values, the documentation and the actual behavior of the WebService might come out of sync?
In the end we can't fully check that and it's up to the user to not do any evil. With the current approach though, a user could return `nil`, which would be illegal for Parameters with non optional type and triggers `fatalError("Could not unwrap opiontal value for \(definition.description) even though it is not an optional. This should have been detected by 'checkNecessity'.")` in your approach, which doesn't quite describe the occurred issue.

**Comment by theMomax** [comment]
> Not sure if I looked at the difference in implementation detailed enough, but wouldn't it be desirable that the default value does not change? In the case where a description of an WebService like OpenAPI creates it, include default values, the documentation and the actual behavior of the WebService might come out of sync?
> In the end we can't fully check that and it's up to the user to not do any evil. With the current approach though, a user could return `nil`, which would be illegal for Parameters with non optional type and triggers `fatalError("Could not unwrap opiontal value for \(definition.description) even though it is not an optional. This should have been detected by 'checkNecessity'.")` in your approach, which doesn't quite describe the occurred issue.

@Supereg I agree with your first point (if OpenAPI really needs a default value). However, I either don't get your last point or I think it is not valid.

The user can only return `nil`, if the `Parameter`'s `Element` is `ExpressibleByNilLiteral`. And then it is a valid value. Note that the `wrappedValue` taken by the `Parameter`'s `init`s is of type `() -> Element`. Only internally `() -> Element?` is used.

**Comment by bauer-andreas** [comment]
> [...] However, I either don't get your last point or I think it is not valid.

> The user can only return nil, if the Parameter's Element is ExpressibleByNilLiteral. And then it is a valid value. Note that the wrappedValue taken by the Parameter's inits is of type () -> Element. Only internally () -> Element? is used.

Ah yeah, I mixed up stuff. Thanks for the clarification 👍 The optional came from the `Type` in `EndpointParameter` I see.

**Review by PSchmiedmayer — APPROVED** [review]
As said in the discussion in #130 I would also prefer non mutable default values. I have merged #130 in this branch. Thanks for the work @theMomax 👍

File: Sources/Apodini/Properties/Parameter.swift [hunk 3]
```diff
@@ -23,8 +23,8 @@ public struct Parameter<Element: Codable>: Property {
     let id: UUID
     let name: String?
     private var element: Element?
-    internal let options: PropertyOptionSet<ParameterOptionNameSpace>
-    internal let defaultValue: Element?
+    internal var options: PropertyOptionSet<ParameterOptionNameSpace>
+    internal var defaultValue: (() -> Element)?
```
Comment (bauer-andreas) [inline Sources/Apodini/Properties/Parameter.swift]: The merge commit seemingly reverted changes done in #106 (Moving to `let`)

**Comment by theMomax** [comment]
I would be fine merging. But keep in mind that the following is still perfectly valid code:
```swift
@Parameter var number = Int.random()
```

**Comment by PSchmiedmayer** [comment]
@theMomax That's a good point 😅
I think we will not be able to change that but at least this PR enable the functionality so you won't be able to easily change it once a `Handler` is initialized for a request and there would be no side effects in case you would (e.g. for a class).

**Pull request merged**